File: examples/directory-server-ts/src/seed.ts
// What a directory seed is made of, and how one is named.
//
// A seed is the JSON `directorySeed` takes: the microdesc consensus, the
// authority certificates that check its signatures, and the microdescriptors
// of every relay in it. The client verifies the consensus against the
// directory authorities pinned in `webtor-core` before installing any of it,
// so nothing here needs to be trusted; what this module checks is only that
// the documents could ever be installed — enough relays, enough signatures
// from authorities the client knows — so a hopeless seed fails here and not
// in every browser.

import { createHash } from 'node:crypto';

export const CONSENSUS_PATH = '/tor/status-vote/current/consensus-microdesc';
/** The `version` the client's `directorySeed` accepts. */
export const SEED_VERSION = 3;
/**
 * Digests per microdescriptor request. Each is a 43-character path segment,
 * so the batch is bounded by the directory's URL length limit.
 */
export const DIGESTS_PER_REQUEST = 90;

/**
 * v3 identity fingerprints of the directory authorities, the set pinned in
 * `crates/webtor-core/src/authority.rs`. The client ignores a signature from
 * anyone else and needs a strict majority of these.
 */
export const AUTHORITY_V3IDENTS: ReadonlySet<string> = new Set([
  '27102BC123E7AF1D4741AE047E160C91ADC76B21', // bastet
  '0232AF901C31A04EE9848595AF9BB7620D4C5B2E', // dannenberg
  'E8A9C45EDE6D711294FADF8E7951F4DE6CA56B58', // dizum
  '70849B868D606BAECFB6128C5E3D782029AA394F', // faravahar
  'ED03BB616EB2F60BEC80151114BB25CEF515B226', // gabelmoo
  '23D15D965BC35114467363C165C4F724B64B4F66', // longclaw
  '49015F787433103580E3B66A1707A00E60F2D15B', // maatuska
  'F533C81CEF0BC0267857C99B2F471ADF249FA232', // moria1
  '2F3DF9CA0E5D36F2685A2DA67184EB8DCB8CBA8C', // tor26
]);
export const CONSENSUS_THRESHOLD = Math.floor(AUTHORITY_V3IDENTS.size / 2) + 1;

/** Floors for a usable consensus, matching what the client refuses to install. */
const MIN_RELAYS_PER_ROLE = 10;
const MIN_HSDIR_RELAYS = 100;

/** A signing authority, as its `directory-signature` line names it. */
export interface Signer {
  /** The authority's v3 identity fingerprint. */
  id: string;
  /** The signing key it used. */
  sk: string;
}

/** What a consensus says about itself, read without checking any of it. */
export interface ConsensusSummary {
  validAfter: Date;
  freshUntil: Date;
  validUntil: Date;
  /** Relays in the consensus. */
  relays: number;
  /** Every relay's microdescriptor digest, deduplicated, in consensus order. */
  digests: string[];
  /** The pinned authorities whose signatures the footer carries. */
  signers: Signer[];
}

/** A seed ready to serve, with everything its manifest says about it. */
export interface Seed {
  /** `<valid-after, compact UTC>-<first 16 hex of the seed's SHA-256>`. */
  name: string;
  /** The seed itself, as `directorySeed` takes it. */
  encoded: string;
  validAfter: Date;
  freshUntil: Date;
  validUntil: Date;
  relays: number;
}

/**
 * Read the consensus's lifetime, its relays and its signers. Throws when the
 * consensus could never be installed: too few relays for a role, or too few
 * signatures from authorities the client knows.
 */
export function summarizeConsensus(consensus: string): ConsensusSummary {
  const validAfter = lifetimeField(consensus, 'valid-after');
  const freshUntil = lifetimeField(consensus, 'fresh-until');
  const validUntil = lifetimeField(consensus, 'valid-until');

  const digests: string[] = [];
  const seen = new Set<string>();
  let relays = 0;
  let middle = 0;
  let hsdir = 0;
  let inFooter = false;
  let current: { digest?: string; flags: Set<string>; bandwidth: number } | null = null;
  const finish = () => {
    if (!current) return;
    relays++;
    if (current.flags.has('HSDir')) hsdir++;
    if (
      current.bandwidth > 0 &&
      current.flags.has('Fast') &&
      current.flags.has('Stable') &&
      current.flags.has('V2Dir')
    ) {
      middle++;
    }
    if (current.digest && !seen.has(current.digest)) {
      seen.add(current.digest);
      digests.push(current.digest);
    }
    current = null;
  };
  const signers = new Map<string, Signer>();

  for (const line of consensus.split('\n')) {
    if (inFooter) {
      if (line.startsWith('directory-signature ')) {
        // `directory-signature [algorithm] <id> <sk>`; the algorithm came
        // later, so the fingerprints are the last two words.
        const words = line.trim().split(/\s+/);
        const id = words.at(-2)?.toUpperCase();
        const sk = words.at(-1)?.toUpperCase();
        if (id && sk && AUTHORITY_V3IDENTS.has(id)) signers.set(`${id}-${sk}`, { id, sk });
      }
      continue;
    }
    if (line.startsWith('directory-footer')) {
      finish();
      inFooter = true;
    } else if (line.startsWith('r ')) {
      finish();
      current = { flags: new Set(), bandwidth: 0 };
    } else if (current && line.startsWith('m ')) {
      current.digest = line.slice(2).trim();
    } else if (current && line.startsWith('s ')) {
      current.flags = new Set(line.slice(2).trim().split(' '));
    } else if (current && line.startsWith('w ')) {
      current.bandwidth = Number(/Bandwidth=(\d+)/.exec(line)?.[1] ?? 0);
    }
  }
  finish();

  if (middle < MIN_RELAYS_PER_ROLE || hsdir < MIN_HSDIR_RELAYS) {
    throw new Error(
      `the consensus has ${relays} relays, ${middle} usable as middles and ${hsdir} HSDirs: too few for a seed`,
    );
  }
  if (signers.size < CONSENSUS_THRESHOLD) {
    throw new Error(
      `the consensus is signed by ${signers.size} known authorities; the client needs ${CONSENSUS_THRESHOLD}`,
    );
  }
  return { validAfter, freshUntil, validUntil, relays, digests, signers: [...signers.values()] };
}

/** The document that carries every certificate the signers used. */
export function certificatesPath(signers: Signer[]): string {
  const segments = signers.map(({ id, sk }) => `${id.toLowerCase()}-${sk.toLowerCase()}`).sort();
  return `/tor/keys/fp-sk/${segments.join('+')}`;
}

/** The documents that together carry every microdescriptor in `digests`. */
export function microdescriptorPaths(digests: string[]): string[] {
  const paths: string[] = [];
  for (let index = 0; index < digests.length; index += DIGESTS_PER_REQUEST) {
    paths.push(`/tor/micro/d/${digests.slice(index, index + DIGESTS_PER_REQUEST).join('-')}`);
  }
  return paths;
}

export function countCertificates(certificates: string): number {
  return certificates.match(/^dir-key-certificate-version /gm)?.length ?? 0;
}

export function countMicrodescriptors(microdescriptors: string): number {
  return microdescriptors.match(/^onion-key$/gm)?.length ?? 0;
}

/** Put the documents together in the shape the client installs, and name it. */
export function assembleSeed(
  consensus: string,
  summary: ConsensusSummary,
  certificates: string,
  microdescriptors: string,
): Seed {
  const encoded = JSON.stringify({
    version: SEED_VERSION,
    consensus,
    certificates,
    microdescriptors,
  });
  const hash = createHash('sha256').update(encoded).digest('hex');
  return {
    name: `${compactUtc(summary.validAfter)}-${hash.slice(0, 16)}`,
    encoded,
    validAfter: summary.validAfter,
    freshUntil: summary.freshUntil,
    validUntil: summary.validUntil,
    relays: summary.relays,
  };
}

/** `2026-09-04T18:00:00Z`: the form the manifest carries. */
export function iso8601(at: Date): string {
  return at.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/** `20260904T180000Z`: the form a file name carries. */
export function compactUtc(at: Date): string {
  return iso8601(at).replaceAll('-', '').replaceAll(':', '');
}

/** `valid-after 2026-09-04 18:00:00` and its siblings, which are UTC. */
function lifetimeField(consensus: string, field: string): Date {
  const match = new RegExp(`^${field} (\\d{4}-\\d{2}-\\d{2}) (\\d{2}:\\d{2}:\\d{2})$`, 'm').exec(
    consensus,
  );
  if (!match) throw new Error(`the consensus has no ${field} line`);
  return new Date(`${match[1]}T${match[2]}Z`);
}

File: crates/webtor-core/src/seed.rs
//! Building a directory seed outside the browser.
//!
//! A `directorySeed` is three Tor documents in one JSON object: a microdesc
//! consensus, the authority certificates that check its signatures, and the
//! microdescriptors it names. The client downloads all three over a bridge
//! circuit when it has to, and that download is the slowest and least reliable
//! part of a bootstrap; a backend with a plain HTTP connection to a directory
//! authority can fetch the same documents in seconds and hand the browser a
//! seed instead.
//!
//! This module is the part of that job which is about Tor rather than about
//! HTTP. The caller does the fetching, from wherever it likes; what is here
//! reads the consensus for the paths to fetch next, checks the result exactly
//! as [`crate::TorClient`] checks a seed on the way in, and encodes it. A seed
//! that passes here is one the client will install; one that fails here would
//! have been rejected in the browser, after being downloaded.

use crate::authority::{authority_certs_path, parse_authority_certs, UncheckedConsensus};
use crate::directory::{
    encode_microdescriptor_digest, process_directory_documents, DirectoryCache,
    DIRECTORY_CACHE_VERSION,
};
use crate::error::{Result, TorError};
use std::collections::HashSet;
use std::time::SystemTime;
use tor_netdoc::doc::netstatus::MdConsensus;

/// The consensus document a directory cache serves, compressed with zlib
/// when the `.z` suffix is added.
pub const CONSENSUS_PATH: &str = "/tor/status-vote/current/consensus-microdesc";

/// How many digests one `/tor/micro/d/` request may name. Each is one
/// 43-character path segment, so this keeps the URL within what a directory
/// cache accepts.
pub const MICRODESCRIPTORS_PER_REQUEST: usize = 90;

/// A consensus that has been parsed but whose signatures are unchecked. It
/// says which certificates check it and nothing else: the relays inside are
/// reachable only through [`Self::verify`].
pub struct UnverifiedConsensus {
    body: String,
    unchecked: UncheckedConsensus,
}

impl UnverifiedConsensus {
    /// Parse `body` as a microdesc consensus that is valid at `now`.
    pub fn parse(body: String, now: SystemTime) -> Result<Self> {
        let unchecked = UncheckedConsensus::parse(&body, now)?;
        Ok(Self { body, unchecked })
    }

    /// The `/tor/keys/fp-sk/…` path that fetches the certificates this
    /// consensus needs, restricted to the pinned authorities. An error means
    /// no pinned authority signed it, which no certificate can repair.
    pub fn certificates_path(&self) -> Result<String> {
        let ids = self.unchecked.required_cert_ids();
        if ids.is_empty() {
            return Err(TorError::ConsensusFetch(
                "Consensus carries no signature from a known directory authority".to_string(),
            ));
        }
        Ok(authority_certs_path(&ids))
    }

    /// Check the signatures against `certificates` — the document the path
    /// from [`Self::certificates_path`] fetched — and open the consensus.
    pub fn verify(self, certificates: String, now: SystemTime) -> Result<VerifiedConsensus> {
        let consensus = self
            .unchecked
            .validate(&parse_authority_certs(&certificates, now)?)?;
        Ok(VerifiedConsensus {
            body: self.body,
            certificates,
            consensus,
        })
    }
}

/// A consensus a strict majority of the pinned authorities signed, ready to
/// name the microdescriptors a seed should carry.
pub struct VerifiedConsensus {
    body: String,
    certificates: String,
    consensus: MdConsensus,
}

impl VerifiedConsensus {
    /// When the consensus became valid.
    pub fn valid_after(&self) -> SystemTime {
        self.consensus.lifetime().valid_after()
    }

    /// When the authorities publish the next consensus. A seed built from this
    /// one is the newest available until then, and stale afterwards.
    pub fn fresh_until(&self) -> SystemTime {
        self.consensus.lifetime().fresh_until()
    }

    /// When the consensus expires. The client refuses a seed past this.
    pub fn valid_until(&self) -> SystemTime {
        self.consensus.lifetime().valid_until()
    }

    /// Every microdescriptor digest the consensus names, each once.
    ///
    /// The client, downloading through one bridge circuit, samples a few
    /// relays per role; a backend fetching from an authority can afford the
    /// whole network, which leaves path selection weighted across all of it.
    pub fn microdescriptor_digests(&self) -> Vec<[u8; 32]> {
        let mut seen = HashSet::new();
        self.consensus
            .relays()
            .iter()
            .map(|router| *router.md_digest())
            .filter(|digest| seen.insert(*digest))
            .collect()
    }

    /// The `/tor/micro/d/…` path that fetches `digests`. Ask for at most
    /// [`MICRODESCRIPTORS_PER_REQUEST`] at a time.
    pub fn microdescriptors_path(digests: &[[u8; 32]]) -> String {
        let segments: Vec<String> = digests.iter().map(encode_microdescriptor_digest).collect();
        format!("/tor/micro/d/{}", segments.join("-"))
    }

    /// Put the consensus, its certificates and `microdescriptors` — the
    /// concatenated bodies the paths from [`Self::microdescriptors_path`]
    /// fetched — through the checks a client applies to a seed, and encode
    /// what passes in the shape `directorySeed` accepts.
    pub fn into_seed(self, microdescriptors: String) -> Result<BuiltSeed> {
        let processed = process_directory_documents(&self.consensus, &microdescriptors)?;
        let lifetime = self.consensus.lifetime();
        let (valid_after, fresh_until, valid_until) = (
            lifetime.valid_after(),
            lifetime.fresh_until(),
            lifetime.valid_until(),
        );
        let encoded = DirectoryCache {
            version: DIRECTORY_CACHE_VERSION,
            consensus: self.body,
            certificates: self.certificates,
            microdescriptors,
        }
        .encode()?;
        Ok(BuiltSeed {
            encoded,
            relay_count: processed.relays.len(),
            middle_count: processed.middle_count,
            hsdir_count: processed.hsdir_count,
            valid_after,
            fresh_until,
            valid_until,
        })
    }
}

/// A seed the client will accept, and what it holds.
#[derive(Clone, Debug)]
pub struct BuiltSeed {
    /// The seed itself, as `directorySeed` takes it.
    pub encoded: String,
    /// Relays with both a consensus entry and a microdescriptor.
    pub relay_count: usize,
    /// Of those, the ones usable as a middle hop.
    pub middle_count: usize,
    /// Of those, the ones on the HSDir ring.
    pub hsdir_count: usize,
    pub valid_after: SystemTime,
    pub fresh_until: SystemTime,
    pub valid_until: SystemTime,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn microdescriptor_paths_are_unpadded_base64_joined_by_dashes() {
        let path = VerifiedConsensus::microdescriptors_path(&[[0; 32], [255; 32]]);
        assert_eq!(
            path,
            "/tor/micro/d/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA-\
             //////////////////////////////////////////8"
        );
    }

    #[test]
    fn a_consensus_that_is_not_a_consensus_is_rejected() {
        let error = UnverifiedConsensus::parse(
            "network-status-version 3\n".to_string(),
            SystemTime::UNIX_EPOCH,
        )
        .err()
        .expect("garbage is not a consensus");
        assert!(error.to_string().contains("consensus"), "{error}");
    }
}
